Documentation/usage/dlv_dap.md

```diff
@@ -17,10 +17,20 @@ The server does not yet accept multiple client connections (--accept-multiclient
 While --continue is not supported, stopOnEntry launch/attach attribute can be used to control if
 execution is resumed at the start of the debug session.
 
+The --connect flag is a special flag that makes the server operate in reverse mode.
```

Comment:
I also think '--connect' is too similar to `dlv connect. For now I just switched to --client-addr, but I am still not sure about use of the term 'client' here either - the client is not any delve DAP client (e.g. editors), but is a "special" client that is listening on the host:port. 

I am more leaning towards hiding this flag from users because this isn't intended for end users.

> Pardon the delay. It took me a couple of tries to understand what's going on and remind myself what RunInTermnial was about. Have I, by any chance, missed a design discussion for this design in some doc or issue?

We had a brief discussion over VC and I meant to defer any detailed discussion until the PR that actually implements the console mode. I hoped the prep-work including this PR and other were small and self-contained, but it seems that's not the case. In the future, if you want to know high-level summary or prefer a design doc, please just request it instead of spending your precious time producing ascii art.

I will update the golang/vscode-go#124 to outline the plan.

> 
> Do I understand correctly that you are trying to achieve the following?
> 
> ```
> +--------+           +---------------+
> │ vscode │           │ dlv-dap @addr │
> +--------+           +---------------+
>             launch req       
>             =======>  
>        runInTerminal req 
>     (dlv dap --connect=addr)
>             <======
> 
> starts . . . . . . . . . . . . . . . . . . . . . in  terminal
>                                                 +------------------------+
>                                                 │$ dlv dap --connect=addr│
>                                                 +------------------------+
>         runInTerminal resp 
>         (dlv pid, shell pid)
>             =======>  
> 
> 
>                       ignores pids b/c has tcp conn
>                                       initialize req
>                                       =======>
>                                       initialize res
>                                       <=======
>                                       launch req
>                                       =======>
>                                                   launches debugger+program
>          initialized event            initialized event
>              <=======  . . . . . . .  <=======
>           launch res                  launch res
>              <=======  . . . . . . .  <=======
>                      forwards everything else  
>                                                   does the actual debugging     
> ```
> 

Yes.

> Some high-level questions come to mind:
> 
> 1. [Spec](https://microsoft.github.io/debug-adapter-protocol/specification#Reverse_Requests_RunInTerminal) says `This is typically used to launch the debuggee in a terminal provided by the client.` So from that and the pid in the args I get an impression that we should be considering starting just the target process in he terminal and then having the original dlv-dap attach to it. Is this something you have considered?  I guess one thing that complicates this is that we will need to start the process suspended (no idea how/if it's possible), so it doesn't do anything useful before he debugger attaches. Not sure if there might also be any permissions issues.

`pid` from RunInTerminal response won't be ignored, that will be used for cleanup when the dlv dap connect mode doesn't connect in time.

Yes. And yes, your guess is correct. 
1) Let me know if there is a simple way of launching the debuggee separately and let it stay in a suspended state until ptraced.
2) If the debuggee is launched with superuser permission, we won't be able to attach a debugger as a normal user. The debugger should be launched as a root as well.
3) That will add one more case to consider in the teardown/stop flow - it's a result of LaunchRequest but debugger will have AttachPid.

> 2. Have you found any good examples of what other adapters are doing (launching debugger or debuggee, etc)?

In our case, delve works as our launcher&debugger.

* python debugger - run launcher (don't get confused by the name 'spawn_debugee').
https://github.com/microsoft/debugpy/blob/2ae7dc96ecc4ca73b119c2056284d203afa10afd/src/debugpy/adapter/launchers.py 
* c++ debugger - run debugger 
https://github.com/microsoft/MIEngine/blob/8456c49b4731971b885c2909f9feaac4bdcd494d/src/MICore/Transports/RunInTerminalTransport.cs
* lldb - run lldb in a terminal agent mode
https://github.com/vadimcn/vscode-lldb/blob/9f05570a5bae4d1e8101593a771e0d2bd26cda87/adapter/src/terminal.rs

Some languages (Dart, ...) launch VM with debugging enabled.

> 3. Is it possible to hook-up a terminal using the new --tty flag?

Maybe? If there is a simple solution that works cross-platform including windows, windows/WSL, ..
I don't know how.

> 4. Will we need this to work with `--accept-multiclient`?

No.